client/src/panel-cnxml-preview.ts

```diff
@@ -177,42 +182,42 @@ export class CnxmlPreviewPanel extends Panel<PanelIncomingMessage, PanelOutgoing
     return activeUri
   }
 
-  private async refreshContents(): Promise<void> {
-    if (this.resourceBinding == null) {
-      return
-    }
+  private async refreshContentsMessage(resource: vscode.Uri): Promise<PanelOutgoingMessage> {
     // TODO: Get resource contents from the language server?
-    const contents = await fs.promises.readFile(this.resourceBinding.fsPath, { encoding: 'utf-8' })
+    const contents = await fs.promises.readFile(resource.fsPath, { encoding: 'utf-8' })
     const doc = new DOMParser().parseFromString(contents)
     tagElementsWithLineNumbers(doc)
     const lineTaggedContents = new XMLSerializer().serializeToString(doc)
-    void this.postMessage({ type: 'refresh', xml: lineTaggedContents })
+    return { type: 'refresh', xml: lineTaggedContents }
   }
 
   isPreviewOf(resource: vscode.Uri | null): boolean {
     return resource?.fsPath === this.resourceBinding?.fsPath
   }
 
   private async rebindToResource(resource: vscode.Uri | null): Promise<void> {
+    const oldBinding = this.resourceBinding
     this.resourceBinding = resource
-    if (resource == null) {
+    if (this.resourceBinding == null) {
       this.panel.webview.html = rawTextHtml('No resource available to preview')
       return
     }
-    await this.rebindWebviewHtmlForResource(resource)
-    await this.refreshContents()
+    const message = await this.refreshContentsMessage(this.resourceBinding)
+    if (oldBinding == null) {
+      const html = await this.reboundWebviewHtmlForResource(this.resourceBinding)
+      const htmlWithMessageInjected = this.injectEnsuredMessages(html, [message])
+      this.panel.webview.html = htmlWithMessageInjected
+    } else {
+      void this.postMessage(message)
+    }
   }
 
-  private async rebindWebviewHtmlForResource(resource: vscode.Uri): Promise<void> {
-    let html = fs.readFileSync(path.join(this.context.resourceRootDir, 'cnxml-preview.html'), 'utf-8')
+  private async reboundWebviewHtmlForResource(resource: vscode.Uri): Promise<string> {
+    let html = await fs.promises.readFile(path.join(this.context.resourceRootDir, 'cnxml-preview.html'), 'utf-8')
     html = addBaseHref(this.panel.webview, resource, html)
     html = fixResourceReferences(this.panel.webview, html, this.context.resourceRootDir)
     html = fixCspSourceReferences(this.panel.webview, html)
-    const reloadComplete = new Promise((resolve, reject) => {
-      this._onDidInnerPanelReload.event(() => { resolve(undefined) })
-    })
-    // This will cause a panel reload, which we must await
-    this.panel.webview.html = html
-    await reloadComplete
+    html = injectCspNonce(html, this.nonce)
```

Comment:
I'm curious, why might we want more than one nonce?

Reply:
I just meant that only one nonce is necessary, but it could be used for more than one attribute. Injecting the nonces later would allow them to all be added at once without needing to store the nonce in `this.nonce`